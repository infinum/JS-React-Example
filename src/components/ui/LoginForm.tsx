import styled from '@emotion/styled';
import { ReactElement } from 'react';
import { useForm } from 'react-hook-form';

const Title = styled.h1``;

const Error = styled.p`
	padding: 4px 8px;
	color: red;
`;

interface ILoginFormProps {
	onSubmit({ email, password }: { email: string; password: string }): void;
	apiErrors?: string;
}

export const LoginForm = ({ onSubmit, apiErrors }: ILoginFormProps): ReactElement => {
	const { handleSubmit, register, errors } = useForm();

	return (
		<>
			<Title>Login</Title>
			<form onSubmit={handleSubmit(onSubmit)}>
				{apiErrors && <Error>{apiErrors}</Error>}

				<input
					name="email"
					ref={register({
						required: 'Required',
						pattern: {
							value: /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i,
							message: 'invalid email address',
						},
					})}
				/>

				{errors.email && <Error>{errors.email.message}</Error>}

				<input
					name="password"
					type="password"
					ref={register({
						required: 'Required',
					})}
				/>

				{errors.password && <Error>{errors.password.message}</Error>}

				<button type="submit">Log in</button>
			</form>
		</>
	);
};
