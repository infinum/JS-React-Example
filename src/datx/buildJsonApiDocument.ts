import { JsonapiDocument, JsonapiResource } from '@/interfaces/Jsonapi';
import { IJsonapiModel, modelToJsonApi } from '@datx/jsonapi';
import uniqWith from 'lodash/uniqWith';

/**
 * This function is used to build a JSON:API document from a model or a collection of models.
 * It will also extract any relationships that are defined on the model in `included` property.
 *
 * @todo Add support for nested relationships?
 *
 * @param model The model or collection of models to build the document from
 * @returns The JSON:API document that can be used for API mocking
 */
export function buildJsonApiDocument<TModel extends IJsonapiModel>(model: TModel | Array<TModel>) {
	const isCollection = Array.isArray(model);

	const data: JsonapiResource<TModel> | Array<JsonapiResource<TModel>> = isCollection
		? model.map((m) => modelToJsonApi(m))
		: modelToJsonApi(model);

	const relationships = (
		Array.isArray(data)
			? data.flatMap((d) => Object.keys(d.relationships || {})).filter(Boolean)
			: Object.keys(data.relationships || {})
	) as Array<keyof TModel>;

	const included = relationships.flatMap((key) => {
		if (isCollection) {
			return model.flatMap((m) => {
				const relationshipModel = m[key] as IJsonapiModel | Array<IJsonapiModel>;

				return Array.isArray(relationshipModel)
					? relationshipModel.map((m) => modelToJsonApi(m))
					: modelToJsonApi(relationshipModel);
			});
		}

		const relationshipModel = model[key] as IJsonapiModel | Array<IJsonapiModel>;

		return Array.isArray(relationshipModel)
			? relationshipModel.map((m) => modelToJsonApi(m))
			: modelToJsonApi(relationshipModel);
	});

	return {
		data,
		...(included.length && { included: uniqWith(included, (a, b) => a.type === b.type && a.id === b.id) }),
	} satisfies JsonapiDocument<TModel>;
}
