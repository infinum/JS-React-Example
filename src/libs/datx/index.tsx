import { createContext, useContext, useMemo } from 'react';
import { Response, IResponse } from 'datx-jsonapi';
import { dequal } from 'dequal/lite';

/**
 * SWR compare
 */
export const compare = (a: any, b: any): boolean => {
	if (a instanceof Response && b instanceof Response) {
		return dequal(a.snapshot.response.data, b.snapshot.response.data);
	}

	return dequal(a, b);
};

const Context = createContext(null);

export const DatxProvider = Context.Provider;

export function useStore() {
	const store = useContext(Context);

	if (!store) {
		throw new Error('useStore must be used within a store Provider.');
	}

	return store;
}

export function useServerSideResponse(data?: IResponse) {
	const store = useStore();

	const response = useMemo(() => {
		if (store && data) {
			return new Response({ data, status: 200 }, store);
		}
	}, [data, store]);

	return response;
}
