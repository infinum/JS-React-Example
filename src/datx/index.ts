import { Collection } from 'datx';
import { jsonapi } from 'datx-jsonapi';

import { Book } from '../models/Book';
import { Author } from '../models/Author';

export class Datx extends jsonapi(Collection) {
	static types = [Book, Author];

	public fetcher = (url: string) => {
		return this.request(`${url}`, 'GET');
	};
}
