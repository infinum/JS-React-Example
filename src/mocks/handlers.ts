import { buildJsonApiDocument } from '@/datx/buildJsonApiDocument';
import { TEST_INVALID_EMAIL, TEST_INVALID_PASSWORD } from '__mocks__/constants';
import { sessionFactory } from '__mocks__/factories';
import { rest } from 'msw';

export const generateTestApiUrl = (path: string) => {
	return process.env.API_TEST_ENDPOINT + path;
};

export const MOCKED_URLS = {
	SessionCurrent: generateTestApiUrl('sessions/current'),
	SessionMe: generateTestApiUrl('sessions/me'),
	Session: generateTestApiUrl('sessions'),
	Flights: generateTestApiUrl('flights'),
} as const;

export const handlers = [
	rest.get(MOCKED_URLS.SessionCurrent, async (_req, res, ctx) => {
		return res(
			ctx.status(401),
			ctx.json({
				errors: [
					{
						status: 'unauthorized',
						code: 'UNAUTHORIZED',
						title: 'Unauthorized',
						detail: 'Must be logged in to perform this action',
					},
				],
			})
		);
	}),
	rest.post(MOCKED_URLS.Session, async (req, res, ctx) => {
		const body = await req.json();
		const emailOverride = body.data?.attributes?.email || undefined;

		return res(
			ctx.status(200),
			ctx.json(
				sessionFactory({
					overrides: {
						email: emailOverride,
					},
				})
			)
		);
	}),
	rest.delete(MOCKED_URLS.SessionMe, async (_req, res, ctx) => {
		return res(ctx.status(204));
	}),
];

export const handlerOverrides = {
	invalidEmailLogin: rest.post(MOCKED_URLS.Session, (_req, res, ctx) => {
		return res(
			ctx.status(422),
			ctx.json({
				errors: [
					{
						status: 'unprocessable entity',
						code: 'UNPROCESSABLE_ENTITY',
						title: 'Unprocessable Entity',
						detail: TEST_INVALID_EMAIL,
						source: { parameter: 'email', pointer: 'data/attributes/email' },
					},
				],
			})
		);
	}),
	invalidPasswordLogin: rest.post(MOCKED_URLS.Session, (_req, res, ctx) => {
		return res(
			ctx.status(422),
			ctx.json({
				errors: [
					{
						status: 'unprocessable entity',
						code: 'UNPROCESSABLE_ENTITY',
						title: 'Unprocessable Entity',
						detail: TEST_INVALID_PASSWORD,
						source: { parameter: 'password', pointer: 'data/attributes/password' },
					},
				],
			})
		);
	}),
	activeCurrentSession: rest.get(MOCKED_URLS.SessionCurrent, async (req, res, ctx) => {
		const session = sessionFactory();

		return res(ctx.status(200), ctx.json(buildJsonApiDocument(session)));
	}),
};
