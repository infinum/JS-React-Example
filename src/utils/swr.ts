import { Model } from 'datx';

/*
  Inspired by https://github.com/lukeed/dequal/blob/master/src/lite.js
*/
export const swrComparator = (a: unknown, b: unknown) => {
  if (a === b) {
    return true;
  }

  if (a && b && a.constructor === b.constructor) {
    let length: number;

    if (Array.isArray(a) && Array.isArray(b)) {
      // eslint-disable-next-line no-cond-assign
      if ((length = a.length) === b.length) {
        while (length-- && swrComparator(a[length], b[length]));
      }
      return length === -1;
    }

    if (a instanceof Model && b instanceof Model) {
      return swrComparator(a.meta.snapshot, b.meta.snapshot);
    }

    if (!a.constructor || typeof a === 'object') {
      length = 0;
      // eslint-disable-next-line guard-for-in, no-restricted-syntax
      for (const key in a as Record<string, unknown>) {
        if (
          Object.prototype.hasOwnProperty.call(a, key) &&
          ++length &&
          !Object.prototype.hasOwnProperty.call(b, key)
        ) {
          return false;
        }
        if (!(key in (b as Record<string, unknown>)) || !swrComparator(a[key], b[key])) {
          return false;
        }
      }

      return Object.keys(b).length === length;
    }
  }

  // eslint-disable-next-line no-self-compare
  return a !== a && b !== b;
};
