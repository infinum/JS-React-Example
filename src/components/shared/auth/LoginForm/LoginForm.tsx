import React, { FC } from 'react';
import { Box, BoxProps, Button } from '@chakra-ui/react';
import { InputField } from '@/components/shared/fields/InputField/InputField';
import { useForm } from 'react-hook-form';
import { setApiErrors } from '@/utils/setApiErrors';
import { Response } from '@datx/jsonapi';
import { useTranslation } from 'next-i18next';
import { login } from '@/mutations/auth';
import { useClient } from '@datx/swr';
import { useSession } from '@/hooks/use-session';

interface IFormValues {
	email: string;
	password: string;
}

export const LoginForm: FC<BoxProps> = () => {
	const { t } = useTranslation('login');
	const {
		register,
		handleSubmit,
		formState: { errors },
		setError,
	} = useForm<IFormValues>();
	const { mutate } = useSession();
	const client = useClient();

	async function onSubmit(formData: IFormValues) {
		try {
			const data = {
				data: {
					type: 'session',
					id: '',
					attributes: {
						...formData,
					},
				},
			};

			await mutate(() => login(client, data), false);
		} catch (errors) {
			if (errors instanceof Response) {
				setApiErrors(errors.error).forEach(({ name, type, message }) => setError(name, { type, message }));
			}
		}
	}

	return (
		<Box as="form" onSubmit={handleSubmit(onSubmit)}>
			<InputField
				label={t('form.email.label')}
				errors={errors}
				{...register('email', { required: t('form.required') })}
			/>
			<InputField
				label={t('form.password.label')}
				errors={errors}
				{...register('password', { required: t('form.required') })}
			/>
			<Button type="submit">{t('form.submit.label')}</Button>
		</Box>
	);
};
