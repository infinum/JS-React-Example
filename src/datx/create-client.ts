import { Collection } from '@datx/core';
import { CachingStrategy, config, ICollectionFetchOpts, IRawResponse } from '@datx/jsonapi';
import { jsonapiSwrClient } from '@datx/swr';

import { Session } from '../models/Session';
import { User } from '../models/User';
import { Flight } from '../models/Flight';
import { apify, deapify } from '@/utils/api-transformers';

export class JsonapiSwrClient extends jsonapiSwrClient(Collection) {
	public static types = [Session, User, Flight];
}

export function createClient() {
	config.baseUrl = (process.env.NEXT_PUBLIC_API_ENDPOINT as string) + '/';
	config.cache = CachingStrategy.NetworkOnly;
	config.fetchReference = fetch;
	config.defaultFetchOptions = {
		headers: {
			Accept: 'application/vnd.api+json',
			'Content-Type': 'application/vnd.api+json',
		},
		credentials: 'include',
	};

	config.transformResponse = (opts: IRawResponse) => {
		console.log('API:', JSON.stringify(opts.data));

		return { ...opts, data: deapify(opts.data) };
	};

	config.transformRequest = (opts: ICollectionFetchOpts) => {
		return { ...opts, data: apify(opts.data) };
	};

	const client = new JsonapiSwrClient();

	return client;
}

export type Client = typeof JsonapiSwrClient;
