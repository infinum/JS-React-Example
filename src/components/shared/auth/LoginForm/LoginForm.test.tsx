import { createClient } from '@/datx/create-client';
import { MOCKED_URLS, handlerOverrides } from '@/mocks/handlers';
import { server } from '@/mocks/server';
import { DatxProvider, useInitialize } from '@datx/swr';
import userEvent from '@testing-library/user-event';
import { axe } from 'jest-axe';
import { FC } from 'react';
import { render, screen, waitFor, waitForRequest } from 'test-utils';
import { LoginForm } from './LoginForm';

const user = userEvent.setup();

describe('LoginForm', () => {
	let UI: FC;

	beforeAll(() => {
		UI = function UI() {
			return (
				<DatxProvider client={useInitialize(createClient)}>
					<LoginForm />
				</DatxProvider>
			);
		};
	});

	it('should matches snapshot', () => {
		const { asFragment } = render(<UI />);

		expect(asFragment()).toMatchSnapshot();
	});

	it('should be accessible', async () => {
		const { container } = render(<UI />);

		const results = await axe(container);

		expect(results).toHaveNoViolations();
	});

	it('should be accessible with form errors', async () => {
		const { container } = render(<UI />);

		await triggerSubmit();

		const results = await axe(container);

		expect(results).toHaveNoViolations();
	});

	it('should display required error when value is invalid', async () => {
		const { container } = render(<UI />);

		await triggerSubmit();

		expect(getErrorMessageElements(container)).toHaveLength(2);
	});

	it('should not display error when value is valid', async () => {
		const { container } = render(<UI />);

		const pendingRequest = waitForRequest.post(MOCKED_URLS.Session);

		const emailValue = 'test@infinum.com';
		const passwordValue = 'password';

		await user.type(screen.getByRole('textbox', { name: /email/i }), emailValue);
		await user.type(screen.getByLabelText(/password/i, { selector: 'input' }), passwordValue);
		await triggerSubmit();

		await waitFor(() => {
			expect(getErrorMessageElements(container)).toHaveLength(0);
		});

		const request = await pendingRequest;

		expect((await request.json())?.data?.attributes).toEqual({
			email: emailValue,
			password: passwordValue,
		});
	});

	it('should display error from API', async () => {
		const { container } = render(<UI />);

		server.use(handlerOverrides.invalidLogin);

		const emailValue = 'test@infinum.com';
		const passwordValue = 'password';

		await user.type(screen.getByRole('textbox', { name: /email/i }), emailValue);
		await user.type(screen.getByLabelText(/password/i, { selector: 'input' }), passwordValue);
		await triggerSubmit();

		await waitFor(() => {
			expect(getErrorMessageElements(container)).toHaveLength(1);
		});
	});
});

const triggerSubmit = async () => {
	const submitButton = screen.queryByRole('button', { name: /submit\.label/i });

	if (submitButton) {
		await user.click(submitButton);
	}
};

const getErrorMessageElements = (container: HTMLElement) => {
	// eslint-disable-next-line testing-library/no-container, testing-library/no-node-access
	return container.querySelectorAll('[aria-live]');
};
