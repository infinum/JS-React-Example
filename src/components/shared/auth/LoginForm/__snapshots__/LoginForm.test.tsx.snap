// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm should matches snapshot 1`] = `
<DocumentFragment>
  <form
    class="chakra-stack css-b5leqj"
    novalidate=""
  >
    <div
      class="chakra-stack css-1au6mu0"
    >
      <div
        class="chakra-form-control css-1kxonj9"
        role="group"
      >
        <label
          class="chakra-form__label css-1ouynl9"
          for="email"
          id="email-label"
        >
          email.label
        </label>
        <input
          class="chakra-input css-9wkudj"
          id="email"
          name="email"
          type="email"
        />
      </div>
      <div
        class="chakra-form-control css-1kxonj9"
        role="group"
      >
        <label
          class="chakra-form__label css-1ouynl9"
          for="password"
          id="password-label"
        >
          password.label
        </label>
        <div
          class="chakra-input__group css-4302v8"
        >
          <div
            class="chakra-input__right-element css-ur97uj"
          >
            <button
              aria-label="Reveal password"
              class="chakra-button css-14grpeb"
              type="button"
            >
              <svg
                aria-hidden="true"
                fill="currentColor"
                focusable="false"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 20 20"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M10 12a2 2 0 100-4 2 2 0 000 4z"
                />
                <path
                  clip-rule="evenodd"
                  d="M.458 10C1.732 5.943 5.522 3 10 3s8.268 2.943 9.542 7c-1.274 4.057-5.064 7-9.542 7S1.732 14.057.458 10zM14 10a4 4 0 11-8 0 4 4 0 018 0z"
                  fill-rule="evenodd"
                />
              </svg>
            </button>
          </div>
          <input
            aria-required="true"
            autocomplete="current-password"
            class="chakra-input css-1rorybn"
            id="password"
            name="password"
            required=""
            type="password"
          />
        </div>
      </div>
    </div>
    <div
      class="chakra-stack css-1oeb4ru"
    >
      <label
        class="chakra-checkbox css-1feyqmz"
        data-checked=""
      >
        <input
          checked=""
          class="chakra-checkbox__input"
          style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
          type="checkbox"
          value=""
        />
        <span
          aria-hidden="true"
          class="chakra-checkbox__control css-57ksmd"
          data-checked=""
        >
          <div
            style="display: flex; align-items: center; justify-content: center; height: 100%; transform: none;"
          >
            <svg
              class="css-1x1o9fj"
              opacity="1"
              stroke-dashoffset="0"
              style="fill: none; stroke-width: 2; stroke: currentColor; stroke-dasharray: 16;"
              viewBox="0 0 12 10"
            >
              <polyline
                points="1.5 6 4.5 9 10.5 1"
              />
            </svg>
          </div>
        </span>
        <span
          class="chakra-checkbox__label css-mq5zsn"
          data-checked=""
        >
          Remember me
        </span>
      </label>
      <button
        class="chakra-button css-13pffa4"
        type="button"
      >
        Forgot password?
      </button>
    </div>
    <div
      class="chakra-stack css-b5leqj"
    >
      <button
        class="chakra-button css-1rnaulk"
        type="submit"
      >
        submit.label
      </button>
    </div>
  </form>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
